src/compiler/checker.ts

```diff
@@ -14040,10 +14040,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return modifiers & MappedTypeModifiers.ExcludeOptional ? -1 : modifiers & MappedTypeModifiers.IncludeOptional ? 1 : 0;
     }
 
+    function getModifiersTypeOptionality(type: Type): number {
+        return type.flags & TypeFlags.Intersection ? Math.max(...map((type as IntersectionType).types, getModifiersTypeOptionality)) :
+            getObjectFlags(type) & ObjectFlags.Mapped ? getMappedTypeOptionality(type as MappedType) :
+            0;
+    }
+
     function getCombinedMappedTypeOptionality(type: MappedType): number {

```

Comment:
If I'm not mistaken this currently falls short here:
```ts
type Obj = {
    a: 1,
    b: 2
};

type Identity<T> = { [K in keyof T]: T[K] };

const mapped: {
    [K in keyof Identity<Partial<Obj>>]: Obj[K]
} = {}

const resolveMapped = <K extends keyof typeof mapped>(key: K) => mapped[key].toString(); // should error but it doesnt
```

Some extra handling of homomorphic mapped types might have to be added here.

Reply:
Yeah, looks like we need to recurse in `getCombinedMappedTypeOptionality`.